Added curl-patch tool

devgianlu opened the pull request:
This tool allows to log all the requests made with `libcurl` to the console.

devgianlu committed:
Added curl-patch

kingosticks commented on curl-patch/Makefile:
```diff
@@ -0,0 +1,21 @@
+.PHONY: clean
+
+clean:
+	rm -f $(DYLIB) $(INPUT)
```
You really want `make clean` to remove curl_patch.c?!

kingosticks commented on curl-patch/Makefile:
```diff
@@ -0,0 +1,21 @@
+DYLIB=curl_patch.dylib
+INPUT=curl_patch.c
+
+curl_patch:
```
`curl_patch: curl_patch.c` ?

kingosticks commented:
Does the libcurl stuff really not end up in the wireshark dump?

devgianlu commented:
It's HTTPS, so it is not very clear what's happening.

On Sun, Jan 6, 2019, 18:00 Nick Steel <notifications@github.com wrote:

> Does the libcurl stuff really not end up in the wireshark dump?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/librespot-org/spotify-analyze/pull/6#issuecomment-451756980>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOI-8fKFmsN9XpHshOXlUB-CvehAMnBSks5vAiubgaJpZM4Zyec0>
> .
>

kingosticks commented on README.md:
```diff
@@ -19,6 +19,15 @@ This will produce a `dump.pcap` in the current directory, which can be analyzed
 The `spotify-dissect` tool is a set of Wireshark dissectors used to analyze packet dumps.
 
+## spotify-curl-patch
+The `spotify-curl-patch` tool sets `CURLOPT_VERBOSE` to `1` logging to the console every request made with `libcurl`.
```
You should explain that this lets you (easily) see the unencrypted requests.

kingosticks commented on README.md:
```diff
@@ -19,6 +19,15 @@ This will produce a `dump.pcap` in the current directory, which can be analyzed
 ## spotify-dissect
 The `spotify-dissect` tool is a set of Wireshark dissectors used to analyze packet dumps.
 
+## spotify-curl-patch
```
Personally I'd have called this something like 'spotify-dump-http'.

plietar commented:
Could this be merged with the existing spotify-dump instead?

devgianlu commented:
~I'd need to create a curl to pcap script.~ I can use `CURLOPT_DEBUGFUNCTION`, I'll give it a try.

devgianlu committed:
Fixes and clarifications

kingosticks commented on dump-http/patch.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
dump-http/patch.sh needs to be executable. It's not.

devgianlu committed:
Made patch.sh executable

devgianlu committed:
Added custom debug function

devgianlu commented:
Writing a `pcap` file is a bit too complicated for my C skills. I'll leave this to someone else.

devgianlu committed:
Fixed name clashing in Wireshark 3.0.0

kingosticks committed:
Display audio key request/response

kingosticks committed:
Decode spirc Frame protobuf responses.

Renamed Frame to SpircFrame since clashes with something pre-existing within Wireshark.
Decode mercury messages with uri starting with 'hm://remote/' using SpircFrame protobuf.

kingosticks committed:
Ignore decrypted packets with unexpected lengths.

I don't know the causes of these packets but this seems better than crashing.

devgianlu committed:
Added `spirc.proto`

sashahilton00 commented:
Can we not write this the the existing pcap file that the mercury stuff gets dumped to? I haven't looked at pcap format, but it was my understanding that it can store multiple dumps of different protocols?

devgianlu commented:
I'm pretty sure that can be done. @kingosticks may be able to do that.

kingosticks commented:
Ideally wouldn't we want both this and the protobuf dumping for a given run? I guess they could each dump to different pcap files and then wireshark can load them both and presumably lineup all the timestamps properly. That'd be cool.

sashahilton00 commented:
I think having it all in one file w/ timestamps would be optimal. 2 seperate files sounds unnecessary. I've been looking into ways to get a realtime capture (on a side note), and will try and add support for FIFO pipes when I get some spare time. Reading over https://wiki.wireshark.org/CaptureSetup/Pipes, it doesn't look like we need to change much to add support to `dump.c`. What I'd ideally like to do is merge dump and dissect into one, so I can just run `./proxy /Applications/Spotify.app/Contents/MacOS/Spotify` and have Wireshark load up and give me a realtime stream of packets, though I appreciate that most people probably want `dump` and `dissect` kept seperate.

kingosticks commented:
I agree but I'd imagine the http stuff will be running in a different thread to the protobuf stuff so we'll have some synchronisation issues when writing to the same file. Maybe they are easy to solve but two files eliminates it entirely.

I've been using it lately where I start the dissector while I'm still capturing and it works fine. Except you do have to click the "reload pcap" button in wireshark everytime you want to see the latest dumped stuff.

sashahilton00 commented:
> I've been using it lately where I start the dissector while I'm still capturing and it works fine. Except you do have to click the "reload pcap" button in wireshark everytime you want to see the latest dumped stuff.

Yeah that does work, but feels a bit clunky. WRT the threads issue, I suspect you are right, though one would have to investigate further to see if that was a significant issue or whether we could resolve it with timestamped events.

k3d3 commented:
Spotify appears to support SSLKEYLOGFILE (I'm not sure if this is from curl or electron, but a lot of traffic appears to be decrypted), so I think that may be the simplest way to get at the traffic.

devgianlu commented:
@k3d3 I've just tried this, but I still see a lot of encrypted traffic.
![image](https://user-images.githubusercontent.com/14827249/64130730-85961380-cdc4-11e9-8820-346ddb6ae776.png)

devgianlu commented:
I've actually managed to decrypt all the traffic, I wasn't able to find it because Spotify uses HTTP2. @k3d3 Thank you a lot!

devgianlu committed:
Dissect event-service messages

devgianlu committed:
Merge branch 'pr/1'

devgianlu committed:
Update from master (#2)

* Update Windows script

* Add patching tool for Spotify macOS

Co-authored-by: Sasha Hilton <sashahilton00@users.noreply.github.com>

devgianlu committed:
Decode spirc frame protobuf (#1)

* dissect: Adjust protocol name to avoid clash with built-in dissector.

Wireshark now includes a Protobuf dissector but it doesn't seem to support
loading .proto files so isn't that useful. Our dissector name clashes with
that built-in one and gives an annoying message on startup.

* Display audio key request/response

* Decode spirc Frame protobuf responses.

Renamed Frame to SpircFrame since clashes with something pre-existing within Wireshark.
Decode mercury messages with uri starting with 'hm://remote/' using SpircFrame protobuf.

* Ignore decrypted packets with unexpected lengths.

I don't know the causes of these packets but this seems better than crashing.

* Update Windows script

Co-authored-by: Nick Steel <nickasteel@gmail.com>
Co-authored-by: Sasha Hilton <sashahilton00@users.noreply.github.com>

devgianlu committed:
Moved event-service stuff to another file

devgianlu committed:
Added 558 event

devgianlu committed:
Added more events

sashahilton00 commented:
I'm going to merge this as is. There were a couple of other enhancements raised in this issue wrt the pcap file stuff and a possible live intercept, but those can be handled separately.
